File: apps/cli/cmd/open.go
package cmd

import (
	"context"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/briandowns/spinner"
	"github.com/apex/cli/internal/chat"
	"github.com/apex/cli/internal/db"
	"github.com/apex/cli/internal/sandbox"
	"github.com/apex/cli/internal/types"
	"github.com/fatih/color"
	"github.com/spf13/cobra"
)

var (
	openPrompt  string
	openGitRepo string
	openStream  bool
)

var openCmd = &cobra.Command{
	Use:   "open <project-id-or-name>",
	Short: "Open a project — interactive or one-shot (-p)",
	Long: `Open an existing project or create a new one on the fly.

If the project doesn't exist and --prompt is provided, a new project
with a sandbox is provisioned automatically.

For throwaway one-shot tasks, use "apex run" instead.`,
	Args: cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		database, err := db.Open(cfg.DBPath)
		if err != nil {
			return fmt.Errorf("open database: %w", err)
		}
		defer database.Close()

		identifier := args[0]

		var projectRow *db.ProjectRow
		var manager *sandbox.Manager
		var created bool

		projectRow, err = resolveProject(database, identifier)
		if err != nil {
			if openPrompt == "" {
				return err
			}
			projectRow, manager, err = createProjectWithSandbox(database, identifier, openGitRepo)
			if err != nil {
				return err
			}
			created = true
		}

		if !created {
			if err := waitForSandbox(database, &projectRow); err != nil {
				return err
			}

			if projectRow.Status == "error" {
				errMsg := "unknown"
				if projectRow.StatusError != nil {
					errMsg = *projectRow.StatusError
				}
				return fmt.Errorf("project sandbox is in error state: %s", errMsg)
			}

			if projectRow.SandboxID == nil || *projectRow.SandboxID == "" {
				return fmt.Errorf("project has no sandbox (status: %s)", projectRow.Status)
			}

			if projectRow.Status == "stopped" {
				color.New(color.FgYellow).Println("Warning: Sandbox is stopped. Attempting to reconnect...")
			}

			manager, err = sandbox.NewManager(cfg.AnthropicKey, cfg.DaytonaKey, cfg.DaytonaURL)
			if err != nil {
				return fmt.Errorf("sandbox manager init: %w", err)
			}

			ctx := context.Background()
			if err := manager.Connect(ctx, *projectRow.SandboxID); err != nil {
				return fmt.Errorf("connect to sandbox: %w", err)
			}
		}

		defer manager.Close()

		if openStream {
			chat.ProgressOut = os.Stderr
		}

		project := rowToProject(projectRow)
		repl := chat.NewREPL(database, manager, project.ID, project)

		if openPrompt != "" {
			return repl.RunSinglePrompt(openPrompt)
		}
		return repl.Run()
	},
}

func init() {
	openCmd.Flags().StringVarP(&openPrompt, "prompt", "p", "", "Send a prompt, run the agent, and exit")
	openCmd.Flags().BoolVarP(&openStream, "stream", "s", false, "Stream task progress to stderr; keeps stdout clean for result output")
	openCmd.Flags().StringVar(&openGitRepo, "git-repo", "", "Git repository URL to clone (used when creating a new project)")
}

// createProjectWithSandbox creates a new project record, provisions a sandbox,
// and returns the project row and a connected manager.
func createProjectWithSandbox(database *db.DB, name, gitRepoURL string) (*db.ProjectRow, *sandbox.Manager, error) {
	userID, err := database.EnsureDefaultUser()
	if err != nil {
		return nil, nil, fmt.Errorf("ensure user: %w", err)
	}

	var gitRepo *string
	if gitRepoURL != "" {
		gitRepo = &gitRepoURL
	}

	projectRow, err := database.CreateProject(userID, name, "", "claude_code", cfg.Snapshot, gitRepo)
	if err != nil {
		return nil, nil, fmt.Errorf("create project: %w", err)
	}

	s := spinner.New(spinner.CharSets[14], 100*time.Millisecond)
	s.Suffix = " Provisioning sandbox..."
	s.Writer = os.Stderr
	s.Start()

	manager, mErr := sandbox.NewManager(cfg.AnthropicKey, cfg.DaytonaKey, cfg.DaytonaURL)
	if mErr != nil {
		s.Stop()
		errMsg := mErr.Error()
		database.UpdateProjectStatus(projectRow.ID, "error", nil, &errMsg)
		return nil, nil, fmt.Errorf("sandbox manager init: %w", mErr)
	}

	ctx := context.Background()
	sandboxID, sErr := manager.CreateSandbox(ctx, projectRow.SandboxSnapshot, name, gitRepoURL)
	if sErr != nil {
		s.Stop()
		errMsg := sErr.Error()
		database.UpdateProjectStatus(projectRow.ID, "error", nil, &errMsg)
		return nil, nil, fmt.Errorf("sandbox provisioning failed: %w", sErr)
	}

	s.Stop()
	database.UpdateProjectStatus(projectRow.ID, "running", &sandboxID, nil)
	color.New(color.FgGreen).Fprintln(chat.ProgressOut, "Sandbox is ready!")

	projectRow, _ = database.GetProject(projectRow.ID)
	return projectRow, manager, nil
}

// waitForSandbox polls the DB until the sandbox reaches a terminal state.
func waitForSandbox(database *db.DB, row **db.ProjectRow) error {
	if (*row).Status != "creating" && (*row).Status != "starting" {
		return nil
	}

	s := spinner.New(spinner.CharSets[14], 100*time.Millisecond)
	s.Suffix = " Waiting for sandbox..."
	s.Writer = os.Stderr
	s.Start()

	for {
		time.Sleep(3 * time.Second)
		updated, err := database.GetProject((*row).ID)
		if err != nil {
			s.Stop()
			return fmt.Errorf("failed to check project status: %w", err)
		}
		if updated.Status == "running" {
			s.Stop()
			*row = updated
			return nil
		}
		if updated.Status == "error" {
			s.Stop()
			errMsg := "unknown error"
			if updated.StatusError != nil {
				errMsg = *updated.StatusError
			}
			return fmt.Errorf("sandbox failed: %s", errMsg)
		}
	}
}

// cleanupEphemeral tears down the sandbox and removes the project record.
func cleanupEphemeral(database *db.DB, manager *sandbox.Manager, row *db.ProjectRow) {
	fmt.Fprintln(chat.ProgressOut)
	s := spinner.New(spinner.CharSets[14], 100*time.Millisecond)
	s.Suffix = " Destroying ephemeral sandbox..."
	s.Writer = os.Stderr
	s.Start()

	if row.SandboxID != nil && *row.SandboxID != "" {
		ctx := context.Background()
		if err := manager.DeleteSandbox(ctx, *row.SandboxID); err != nil {
			s.Stop()
			color.New(color.FgYellow).Fprintf(chat.ProgressOut, "Warning: failed to delete sandbox: %v\n", err)
		}
	}

	database.DeleteProject(row.ID)
	s.Stop()
	color.New(color.FgGreen).Fprintln(chat.ProgressOut, "Ephemeral sandbox destroyed.")
}

// resolveProject finds a project by exact ID or by name prefix match.
func resolveProject(database *db.DB, identifier string) (*db.ProjectRow, error) {
	project, err := database.GetProject(identifier)
	if err == nil {
		return project, nil
	}

	project, err = database.GetProjectByName(identifier)
	if err == nil && project != nil {
		return project, nil
	}

	projects, listErr := database.ListProjects(db.DefaultUserID)
	if listErr != nil {
		return nil, fmt.Errorf("failed to list projects: %w", listErr)
	}

	var matches []db.ProjectRow
	lower := strings.ToLower(identifier)
	for _, p := range projects {
		if strings.HasPrefix(strings.ToLower(p.Name), lower) ||
			strings.HasPrefix(p.ID, identifier) {
			matches = append(matches, p)
		}
	}

	if len(matches) == 1 {
		return &matches[0], nil
	}
	if len(matches) > 1 {
		names := make([]string, len(matches))
		for i, m := range matches {
			shortID := m.ID
			if len(shortID) > 8 {
				shortID = shortID[:8]
			}
			names[i] = fmt.Sprintf("  %s (%s)", m.Name, shortID)
		}
		return nil, fmt.Errorf("ambiguous project name %q, matches:\n%s", identifier, strings.Join(names, "\n"))
	}

	return nil, fmt.Errorf("project not found: %s", identifier)
}

func rowToProject(row *db.ProjectRow) *types.Project {
	return &types.Project{
		ID:          row.ID,
		UserID:      row.UserID,
		Name:        row.Name,
		Description: row.Description,
		SandboxID:   row.SandboxID,
		Status:      row.Status,
		StatusError: row.StatusError,
		AgentType:   row.AgentType,
		GitRepo:     row.GitRepo,
		CreatedAt:   row.CreatedAt,
		UpdatedAt:   row.UpdatedAt,
	}
}

File: apps/cli/src/commands/open.ts
import { Command } from 'commander';
import chalk from 'chalk';
import { randomBytes } from 'crypto';
import { DatabaseManager } from '../database/bun-sqlite.js';
import { MockSandboxManager } from '../sandbox/mock.js';
import { ThreadManager } from '../thread/index.js';
import { configManager } from '../config/index.js';
import type { Project } from '../types/index.js';

interface OpenOptions {
  prompt?: string;
  stream?: boolean;
  gitRepo?: string;
}

export function createOpenCommand(): Command {
  const command = new Command('open')
    .description('Open a project — interactive or one-shot (-p)')
    .argument('<project-id-or-name>', 'Project ID or name to open')
    .option('-p, --prompt <prompt>', 'Send a prompt, run the agent, and exit')
    .option('-s, --stream', 'Stream task progress to stderr; keeps stdout clean for result output', false)
    .option('--git-repo <url>', 'Git repository URL to clone (used when creating a new project)')
    .action(async (identifier: string, options: OpenOptions) => {
      const config = configManager.config;
      
      if (!config.anthropicApiKey || !config.daytonaApiKey) {
        console.error(chalk.red('API keys not configured. Run "apex configure" to set them up.'));
        console.error(chalk.gray(`Database: ${config.dbPath}`));
        process.exit(1);
      }

      const db = new DatabaseManager(config.dbPath);
      const sandboxManager = new MockSandboxManager();
      const threadManager = new ThreadManager(db, sandboxManager);

      try {
        let project = resolveProject(db, identifier);
        let created = false;

        if (!project) {
          // Project doesn't exist
          if (!options.prompt) {
            console.error(chalk.red(`Project "${identifier}" not found`));
            console.log(chalk.gray('Available projects:'));
            const projects = db.listProjects();
            if (projects.length === 0) {
              console.log(chalk.gray('  No projects exist. Create one with: apex create <name>'));
            } else {
              for (const p of projects) {
                const shortId = p.id.length > 8 ? p.id.slice(0, 8) : p.id;
                console.log(chalk.gray(`  ${p.name} (${shortId})`));
              }
            }
            process.exit(1);
          }

          // Create project on-the-fly for one-shot execution
          console.log(chalk.cyan(`🚀 Creating project "${identifier}" on-the-fly...`));
          
          const user = db.getDefaultUser();
          const projectId = `proj-${randomBytes(8).toString('hex')}`;
          
          project = db.createProject({
            id: projectId,
            userId: user.id,
            name: identifier,
            provider: config.defaultProvider,
            status: 'creating',
            agentType: config.defaultAgentType,
            gitRepo: options.gitRepo,
            agentConfig: {},
          });

          // Create sandbox
          const sandboxId = await sandboxManager.createSandbox(project);
          
          // Update project with sandbox ID
          db.updateProject(project.id, { 
            sandboxId, 
            status: 'running' 
          });
          project.sandboxId = sandboxId;
          project.status = 'running';
          created = true;
        }

        if (!created) {
          // Check project status and connect to existing sandbox
          await waitForSandbox(db, project, sandboxManager);
          
          if (project.status === 'error') {
            console.error(chalk.red(`Project sandbox is in error state`));
            process.exit(1);
          }

          if (!project.sandboxId) {
            console.error(chalk.red(`Project has no sandbox (status: ${project.status})`));
            process.exit(1);
          }

          if (project.status === 'stopped') {
            console.log(chalk.yellow('⚠️  Sandbox is stopped. Attempting to reconnect...'));
          }

          await sandboxManager.connectToSandbox(project.sandboxId);
        }

        // Configure output for streaming or clean stdout
        if (options.stream && !options.prompt) {
          // Stream progress to stderr for interactive mode
          sandboxManager.setCallbacks({
            onProgress: (msg) => console.error(chalk.gray(msg)),
            onStatusChange: (status) => console.error(chalk.gray(`Status: ${status}`)),
          });
        } else if (options.prompt && !options.stream) {
          // Disable progress output for clean stdout in one-shot mode
          sandboxManager.setCallbacks({
            onProgress: () => {},
            onStatusChange: () => {},
          });
        }

        if (options.prompt) {
          // One-shot execution
          await threadManager.startThread(project, {
            interactive: false,
            oneShot: true,
            initialPrompt: options.prompt,
          });
        } else {
          // Interactive session
          console.log(chalk.green(`✓ Connected to project "${project.name}"`));
          await threadManager.startThread(project, {
            interactive: true,
          });
        }

      } catch (error) {
        console.error(chalk.red('Operation failed:'), (error as Error).message);
        process.exit(1);
      } finally {
        threadManager.cleanup();
        sandboxManager.disconnect();
        db.close();
      }
    });

  return command;
}

function resolveProject(db: DatabaseManager, identifier: string): Project | null {
  // First try exact ID match
  let project = db.getProject(identifier);
  if (project) return project;

  // Then try name match
  const projects = db.listProjects();
  
  // Exact name match
  project = projects.find(p => p.name.toLowerCase() === identifier.toLowerCase()) || null;
  if (project) return project;

  // Prefix matches
  const lowerIdentifier = identifier.toLowerCase();
  const matches = projects.filter(p => 
    p.name.toLowerCase().startsWith(lowerIdentifier) ||
    p.id.startsWith(identifier)
  );

  if (matches.length === 1) {
    return matches[0];
  }

  if (matches.length > 1) {
    console.error(chalk.red(`Ambiguous project name "${identifier}", matches:`));
    for (const match of matches) {
      const shortId = match.id.length > 8 ? match.id.slice(0, 8) : match.id;
      console.error(chalk.gray(`  ${match.name} (${shortId})`));
    }
    process.exit(1);
  }

  return null;
}

async function waitForSandbox(
  db: DatabaseManager,
  project: Project,
  sandboxManager: MockSandboxManager
): Promise<void> {
  if (project.status !== 'creating' && project.status !== 'stopped') {
    return;
  }

  console.log(chalk.cyan('⏳ Waiting for sandbox...'));
  
  // Simple polling - in a real implementation you might want exponential backoff
  let attempts = 0;
  const maxAttempts = 60; // 3 minutes at 3 second intervals
  
  while (attempts < maxAttempts) {
    await new Promise(resolve => setTimeout(resolve, 3000));
    
    const updated = db.getProject(project.id);
    if (!updated) {
      throw new Error('Project was deleted while waiting');
    }
    
    if (updated.status === 'running') {
      Object.assign(project, updated);
      console.log(chalk.green('✓ Sandbox is ready'));
      return;
    }
    
    if (updated.status === 'error') {
      throw new Error('Sandbox failed to start');
    }
    
    attempts++;
  }
  
  throw new Error('Timeout waiting for sandbox to be ready');
}